File: ArxProjects/cadtest1/NetCADTest/tools.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

#if ZRX
using ZWCAD;
using ZwSoft.ZwCAD.EditorInput;
using ZwSoft.ZwCAD.ApplicationServices;
using ZwSoft.ZwCAD.Runtime;
using ZwSoft.ZwCAD.DatabaseServices;
using ZwSoft.ZwCAD.Geometry;
using ZwSoft.ZwCAD.Windows;
using TransactionManager = ZwSoft.ZwCAD.DatabaseServices.TransactionManager;
#else
using Autodesk.AutoCAD.EditorInput;
using Autodesk.AutoCAD.ApplicationServices;
using Autodesk.AutoCAD.Runtime;
using Autodesk.AutoCAD.DatabaseServices;
using Autodesk.AutoCAD.Geometry;
using Autodesk.AutoCAD.Windows;
using Autodesk.AutoCAD.Internal;
using TransactionManager = Autodesk.AutoCAD.DatabaseServices.TransactionManager;
#endif

[assembly: CommandClass(typeof(NetCADTest.Tools))]

namespace NetCADTest
{
    public class Tools
    {
        // 获取实体的包围盒
        [CommandMethod("getGeoExtentsOfThisEntity")]
        public void getGeoExtentsOfThisEntity()
        {
            Database db = HostApplicationServices.WorkingDatabase;
            Document doc = Application.DocumentManager.MdiActiveDocument;
            Editor ed = doc.Editor;

            PromptEntityOptions entOpts = new PromptEntityOptions("选择对象: ");
            PromptEntityResult entRes = entRes = ed.GetEntity(entOpts);

            if (entRes.Status == PromptStatus.OK)
            {
                ObjectId objId = entRes.ObjectId;

                using (Transaction trans = db.TransactionManager.StartTransaction())
                {
                    Entity entity = (Entity)trans.GetObject(objId, OpenMode.ForRead, true);

                    Extents3d ext = entity.GeometricExtents;
                    ed.WriteMessage("\next.minPoint({0}, {1}, {2}) ext.maxPoint({3}, {4}, {5})",
                       ext.MinPoint.X, ext.MinPoint.Y, ext.MinPoint.Z, ext.MaxPoint.X, ext.MaxPoint.Y, ext.MaxPoint.Z);

                    trans.Commit();
                }
            }
        }

        // 获取两个实体的交点
        [CommandMethod("getIntersectionOfTwoEntities")]
        public void getIntersectionOfTwoEntities()
        {
            Database db = HostApplicationServices.WorkingDatabase;
            Document doc = Application.DocumentManager.MdiActiveDocument;
            Editor ed = doc.Editor;

            PromptEntityOptions entOpts1 = new PromptEntityOptions("Select Entity 1:");
            PromptEntityResult entRes1  = ed.GetEntity(entOpts1);

            PromptEntityOptions entOpts2 = new PromptEntityOptions("Select Entity 2:");
            PromptEntityResult entRes2 = ed.GetEntity(entOpts2);

            if (entRes1.Status == PromptStatus.OK && entRes2.Status == PromptStatus.OK)
            {
                ObjectId objId1 = entRes1.ObjectId;
                ObjectId objId2 = entRes2.ObjectId;

                using (Transaction trans = db.TransactionManager.StartTransaction())
                {
                    Entity entity1 = (Entity)trans.GetObject(objId1, OpenMode.ForRead, true);
                    Entity entity2 = (Entity)trans.GetObject(objId2, OpenMode.ForRead, true);

                    Point3dCollection points = new Point3dCollection();
                    entity1.IntersectWith(entity2, Intersect.OnBothOperands, points, IntPtr.Zero, IntPtr.Zero);

                    ed.WriteMessage("\npoints length: {0}", points.Count);
                    for (int i = 0; i < points.Count; i++)
                    {
                        ed.WriteMessage("\nptInsec[{0}]: ({1}, {2}, {3})", i, points[i].X, points[i].Y, points[i].Z);
                    }

                    trans.Commit();
                }
            }
        }
    }
}

File: ArxProjects/cadtest1/VBCADTest/VBCADTest/Tools.vb
Imports System
#If ZRX Then
Imports ZwSoft.ZwCAD.Runtime
Imports ZwSoft.ZwCAD.ApplicationServices
Imports ZwSoft.ZwCAD.DatabaseServices
Imports ZwSoft.ZwCAD.Geometry
Imports ZwSoft.ZwCAD.EditorInput
#Else
Imports Autodesk.AutoCAD.Runtime
Imports Autodesk.AutoCAD.ApplicationServices
Imports Autodesk.AutoCAD.DatabaseServices
Imports Autodesk.AutoCAD.Geometry
Imports Autodesk.AutoCAD.EditorInput
#End If

<Assembly: CommandClass(GetType(VBCADTest.Tools))>

Namespace VBCADTest
    Public Class Tools
        '获取实体的包围盒
        <CommandMethod("TestVBGroup", "getGeoExtentsOfThisEntity", CommandFlags.Modal)>
        Public Sub getGeoExtentsOfThisEntity()
            Dim db As Database = HostApplicationServices.WorkingDatabase
            Dim doc As Document = Application.DocumentManager.MdiActiveDocument
            Dim ed As Editor = doc.Editor

            Dim entOpts As PromptEntityOptions = New PromptEntityOptions("选择对象: ")
            Dim entRes As PromptEntityResult = ed.GetEntity(entOpts)

            If entRes.Status = PromptStatus.OK Then
                Dim objId As ObjectId = entRes.ObjectId

                Using trans As Transaction = db.TransactionManager.StartTransaction()
                    Dim entity As Entity = trans.GetObject(objId, OpenMode.ForRead, True)

                    Dim ext As Extents3d = entity.GeometricExtents
                    ed.WriteMessage(vbLf + "ext.minPoint({0}, {1}, {2}) ext.maxPoint({3}, {4}, {5})",
                       ext.MinPoint.X, ext.MinPoint.Y, ext.MinPoint.Z, ext.MaxPoint.X, ext.MaxPoint.Y, ext.MaxPoint.Z)

                    trans.Commit()
                End Using
            End If
        End Sub

        '获取两个实体的交点
        <CommandMethod("TestVBGroup", "getIntersectionOfTwoEntities", CommandFlags.Modal)>
        Public Sub getIntersectionOfTwoEntities()
            Dim db As Database = HostApplicationServices.WorkingDatabase
            Dim doc As Document = Application.DocumentManager.MdiActiveDocument
            Dim ed As Editor = doc.Editor

            Dim entOpts1 As PromptEntityOptions = New PromptEntityOptions("Select Entity 1:")
            Dim entRes1 As PromptEntityResult = ed.GetEntity(entOpts1)

            Dim entOpts2 As PromptEntityOptions = New PromptEntityOptions("Select Entity 2:")
            Dim entRes2 As PromptEntityResult = ed.GetEntity(entOpts2)

            If entRes1.Status = PromptStatus.OK And entRes2.Status = PromptStatus.OK Then
                Dim objId1 As ObjectId = entRes1.ObjectId
                Dim objId2 As ObjectId = entRes2.ObjectId

                Using trans As Transaction = db.TransactionManager.StartTransaction()
                    Dim entity1 As Entity = trans.GetObject(objId1, OpenMode.ForRead, True)
                    Dim entity2 As Entity = trans.GetObject(objId2, OpenMode.ForRead, True)

                    Dim points As Point3dCollection = New Point3dCollection()
                    entity1.IntersectWith(entity2, Intersect.OnBothOperands, points, IntPtr.Zero, IntPtr.Zero)

                    ed.WriteMessage(vbLf + "points length: {0}", points.Count)

                    For i As Integer = 0 To points.Count - 1
                        ed.WriteMessage(vbLf + "ptInsec[{0}]: ({1}, {2}, {3})", i, points(i).X, points(i).Y, points(i).Z)
                    Next

                    trans.Commit()
                End Using
            End If
        End Sub
    End Class
End Namespace
